Source/Particles/Deposition/SharedDepositionUtils.H

```diff
@@ -0,0 +1,281 @@
+    }
+
+    // --- Compute shape factors
+    Compute_shape_factor< depos_order > const compute_shape_factor;
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ) || defined(WARPX_DIM_3D)
+
+    // x direction
+    // Get particle position after 1/2 push back in position
+#if defined(WARPX_DIM_RZ)
+    // Keep these double to avoid bug in single precision
+    const double xmid = (rpmid - xmin)*dxi;
+#else
+    const double xmid = ((xp - xmin) + relative_time*vx)*dxi;
+#endif
+    // j_j[xyz] leftmost grid point in x that the particle touches for the centering of each current
+    // sx_j[xyz] shape factor along x for the centering of each current
+    // There are only two possible centerings, node or cell centered, so at most only two shape factor
+    // arrays will be needed.
+    // Keep these double to avoid bug in single precision
+    double sx_node[depos_order + 1] = {0.};
+    double sx_cell[depos_order + 1] = {0.};
+    int j_node = 0;
+    int j_cell = 0;
+    if (j_type[0] == NODE) {
+        j_node = compute_shape_factor(sx_node, xmid);
+    }
+    if (j_type[0] == CELL) {
+        j_cell = compute_shape_factor(sx_cell, xmid - 0.5);
+    }
+
+    amrex::Real sx_j[depos_order + 1] = {0._rt};
+    for (int ix=0; ix<=depos_order; ix++)
+    {
+        sx_j[ix] = ((j_type[0] == NODE) ? amrex::Real(sx_node[ix]) : amrex::Real(sx_cell[ix]));
+    }
+
+    int const j_j = ((j_type[0] == NODE) ? j_node : j_cell);
+#endif //AMREX_SPACEDIM >= 2
+
+#if defined(WARPX_DIM_3D)
+    // y direction
+    // Keep these double to avoid bug in single precision
+    const double ymid = ((yp - ymin) + relative_time*vy)*dyi;
+    double sy_node[depos_order + 1] = {0.};
+    double sy_cell[depos_order + 1] = {0.};
+    int k_node = 0;
+    int k_cell = 0;
+    if (j_type[1] == NODE) {
+        k_node = compute_shape_factor(sy_node, ymid);
+    }
+    if (j_type[1] == CELL) {
+        k_cell = compute_shape_factor(sy_cell, ymid - 0.5);
+    }
+    amrex::Real sy_j[depos_order + 1] = {0._rt};
+    for (int iy=0; iy<=depos_order; iy++)
+    {
+        sy_j[iy] = ((j_type[1] == NODE) ? amrex::Real(sy_node[iy]) : amrex::Real(sy_cell[iy]));
+    }
+    int const k_j = ((j_type[1] == NODE) ? k_node : k_cell);
+#endif
+
+    // z direction
+    // Keep these double to avoid bug in single precision
+    const double zmid = ((zp - zmin) + relative_time*vz)*dzi;
+    double sz_node[depos_order + 1] = {0.};
+    double sz_cell[depos_order + 1] = {0.};
+    int l_node = 0;
+    int l_cell = 0;
+    if (j_type[zdir] == NODE) {
+        l_node = compute_shape_factor(sz_node, zmid);
+    }
+    if (j_type[zdir] == CELL) {
+        l_cell = compute_shape_factor(sz_cell, zmid - 0.5);
+    }
+    amrex::Real sz_j[depos_order + 1] = {0._rt};
+    for (int iz=0; iz<=depos_order; iz++)
+    {
+        sz_j[iz] = ((j_type[zdir] == NODE) ? amrex::Real(sz_node[iz]) : amrex::Real(sz_cell[iz]));
+    }
+    int const l_j = ((j_type[zdir] == NODE) ? l_node : l_cell);
```

Comment:
```suggestion
    const double zmid = ((zp - zmin) + relative_time*vz)*dzi;
    double sz_j[depos_order + 1] = {0.};
    int const l_j = compute_shape_factor(sz_j, (j_type[zdir] == NODE) ? zmid : zmid - 0.5);
```
Same for x and y direction

Reply:
Looks good! Tested in 3D, 2D, and RZ, then committed this.

Reply:
Actually, after further testing, this breaks things for single precision. 
amrex::Gpu::Atomic::AddNoRet requires that sz_j be an amrex::Real  while compute_shape_factor requires sz_j be a double.
The previous version of the code called compute_shape_factor on the double sz_cell, and AddNoRet on the amrex::Real sz_j. so there was no issue. I saw two obvious ways to fix this. The first is to cast sz_j when we call AddNoRet, and the second is to add a for loop after compute_shape_factor to initialize a new sz_j of type amrex::Real. I tested both methods and found them slower (results below). I'm therefore reverting to previous version of code. If anyone finds a better way to keep the simplicity of the proposed changes without sacrificing runtime, will reconsider.
outCast:Total Time                     : 2.67980957
outFor:Total Time                     : 2.624755859
outDev:Total Time                     : 2.297363281

Reply:
The general solution here is to avoid `double` that we use right now and instead calculate the position deltas from positions that were reduced by the offset of the current box (or even cell).

The only reason we introduced double here as a work-around is that our positions are global and frequently run into precision issues z = z + dz for particles in moving window simulations.